lib/record_store/provider/oracle_cloud_dns.rb

```diff
@@ -0,0 +1,215 @@
+
+        client.get_zone_records(
+          zone,
+          rtype: record.type,
+          zone_version: client.get_zone(zone).data.version,
+          domain: record_fqdn,
+        ).each do |response|
+          response.data.items.each do |r|
+            record_hash = r.record_hash
+          end
+        end
+
+        patch_remove_record = [
+          OCI::Dns::Models::RecordOperation.new(
+            domain: record_fqdn,
+            record_hash: record_hash,
+            rtype: record.type,
+            ttl: record.ttl,
+            rdata: record_rdata,
+            operation: 'REMOVE',
+          ),
+        ]
+
+        client.patch_zone_records(
+          zone,
+          OCI::Dns::Models::PatchZoneRecordsDetails.new(items: patch_remove_record)
+        )
+      end
+
+      # Updates an existing record in the zone. It is expected this call modifies external state.
+      #
+      # Arguments:
+      # id - provider specific ID of record to update
+      # record - a kind of `Record` which the record with `id` should be updated to
+      def update(id, record, zone)
+        record_fqdn = record.fqdn.sub(/\.$/, '')
+        record_rdata = retrieve_rdata_based_on_record_type(record)
+
+        # Retrieve all records that you want to keep before it's updated because it will overwrite
+        all_records = []
+        client.get_zone_records(zone, zone_version: zone_version(zone)).each do |response|
+          all_records += response.data.items
+        end
+
+        update_zone_record_items = all_records.map do |r|
+          OCI::Dns::Models::RecordDetails.new(
+            domain: r.domain,
+            ttl: r.ttl,
+            rtype: r.rtype,
+            recordHash: r.record_hash,
+            rdata: r.rdata,
+            rrsetVersion: r.rrset_version,
+            isProtected: true
+          )
+        end
+        update_zone_record_items << OCI::Dns::Models::RecordDetails.new(
+          domain: record_fqdn,
+          ttl: record.ttl,
+          rtype: record.type,
+          rdata: record_rdata
+        )
+        update_zone_record_items.delete_if { |r| id == r.record_hash }
+
+        client.update_zone_records(
+          zone,
+          OCI::Dns::Models::UpdateZoneRecordsDetails.new(items: update_zone_record_items)
+        )
+      end
+
+      def secrets
+        super.fetch('oracle_cloud_dns')
+      end
+
+      def build_from_api(api_record)
+        fqdn = Record.ensure_ends_with_dot(api_record.domain)
+
+        record_type = api_record.rtype
+        return if record_type == 'SOA'
+
+        record = {
+          ttl: api_record.ttl,
+          fqdn: fqdn.downcase,
+          record_id: api_record.record_hash,
+        }
+        case record_type
+        when 'A', 'AAAA'
+          record.merge!(address: api_record.rdata)
+        when 'ALIAS'
+          record.merge!(alias: api_record.rdata)
+        when 'CAA'
+          flags, tag, value = api_record.rdata.split(' ')
+
+          record.merge!(
```

Comment:
Feels unnecessarily un-DRY.  Invites typos in the repeated hash name,`record`.

Reply:
Rubocop usually recommends to avoid using `merge!` except when merging from a hash currently stored in a variable: https://www.rubydoc.info/github/bbatsov/RuboCop/RuboCop/Cop/Performance/RedundantMerge

Reply:
Hmm, considering consistency, NS1 also needs to be updated if I update above.

Reply:
Since there are some that have more than 2 parameters, `merge!`'d be better option here and didn't get any warnings by Rubocop.

Reply:
True, but this can be done separately from this as another PR later for consistency clean-up
Anyway, for the single lines, I would strongly recommend getting rid of the `merge!`, for the multiline ones, I'm also more inclined to the use of direct assignments (as I like the justification that you anyway need to specify things one by one, so might as well do it with assignments) but I'm not feeling as strongly about it. In both cases anyway, it works as-is, so you decide what you want to do!